### PR TITLE
htlcswitch/link: add hash resolver

### PR DESCRIPTION
**Pull request by sangaman**
This PR introduces an optional feature to resolve a hash via an external service. If we receive a payment for which we are the final destination, and yet `LookupInvoice` returns an error indicating an unknown hash, we can attempt to retrieve the preimage for this hash from the external hash resolver service before failing the htlc. This allows for payment hashes and preimages to be created, stored, and retrieved independently from lnd.

One key use case for this functionality is cross chain swaps as described [here](https://github.com/ExchangeUnion/Lightning-Swap-PoC-v2/blob/master/Concept.md). In short, when an unrecognized payment hash from lnd is passed to the hash resolver, the hash resolver service can then attempt to fetch the preimage for that hash by making a lightning payment on a separate lnd instance (using a separate chain/network) using the same hash. A successful payment will reveal the preimage, which can then be returned to the first lnd instance which queried the hash resolver service.

To minimize the changes to existing lnd code, this PR uses a separate configuration file for new options pertaining to querying the resolver service. This can be changed to use the main configuration file, perhaps in a new `[resolver]` section.

### CODE REVIEW

**Comment by offerm** [comment]
Travis [test](https://travis-ci.org/lightningnetwork/lnd/jobs/423155623) failed due to #1592

File: Gopkg.toml [hunk 1]
```diff
@@ -109,3 +109,7 @@
 [prune]
   go-tests = true
   unused-packages = true
+
+[[constraint]]
+  branch = "master"
+  name = "github.com/ExchangeUnion/swap-resolver"
```
Comment (Roasbeef) [inline Gopkg.toml]: We're actively making efforts to _reduce_ the set of dependancies we pull into the project. Why can't this be added as a _purely_ RPC feature?

Comment (offerm) [reply Gopkg.toml]: Resolved by latest commit. 
Dependency removed.

**Comment by offerm** [comment]
@Roasbeef  @halseth I removed the dependency on the external package from this PR.

The new RPC call is now part of LND's rpc.proto. 

I structured the solution in a way that by default the new functionality can be compiled out so this change brings no risk for users which does not need it.

Can you have a look please?

**Comment by offerm** [comment]
@halseth @Roasbeef can you please have a quick look and let me know what can I do to help you digest this one.

I can make more changes to reduce the impact on non-users of this feature to zero. 
Will be happy to get some feedback before I make these changes.

Thanks!

**Comment by cfromknecht** [comment]
@sangaman see #2022

**Pull request closed without merging**